### ISSUE TITLE
Nested StructBlocks in a StreamBlock are not showing the ValidationError in the field

### ISSUE THREAD

**Issue by jkevingutierrez** [wagtail/wagtail#7248]
Found a bug? Please fill out the sections below. 👍

### Issue Summary

In Wagtail 2.13.1, having a required field in a nested StructBlock shows the message "The page could not be saved due to validation errors" at the top of the page, but it doesn't show the message "This field is required." in the failing field.

Looks like the issue is happening with any ValidationError, not just the required error. I also tested it with a custom RegexValidator and it didn't show the error in the field

![image](https://user-images.githubusercontent.com/6220055/121612387-ecc40b80-ca1f-11eb-875a-a7510b127c76.png)

![image](https://user-images.githubusercontent.com/6220055/121612373-e5046700-ca1f-11eb-8d9f-c87e8d6990ec.png)

### Steps to Reproduce

1. Create a StructBlock with some required fields. eg:
```
class CopyBlock(blocks.StructBlock):
    text = blocks.RichTextBlock(required=True)
```
2. Create a StructBlock that will accept other struct blocks using a StreamBlock. eg:
```
class GridLayoutBlock(blocks.StructBlock):
    body = blocks.StreamBlock(
        [
            ('copy', CopyBlock()),
        ]
    )
```
3. Following the example, add GridLayoutBlock to some page StreamField
```
class CustomPage(Page):
    body = StreamField(
        [
            ('grid', GridLayoutBlock()),
        ]
    )
```
4. Create a CustomPage, add a GridLayoutBlock and inside that add a CopyBlock filling the text field.
5. Refresh the page and delete the text field data. Try to save again. It will show the "The page could not be saved due to validation errors" message at the top of the page, but it won't show the message "This field is required." in the field

Also, it doesn't "count" that error in the CONTENT tab.

![image](https://user-images.githubusercontent.com/6220055/121612529-29900280-ca20-11eb-9734-2b1fa0aecd5e.png)

Expected behaviour:

It should show the "This field is required" message in the empty field.

![image](https://user-images.githubusercontent.com/6220055/121612669-770c6f80-ca20-11eb-8531-04d1018a465d.png)

And count the error in the CONTENT tab

![image](https://user-images.githubusercontent.com/6220055/121612624-61974580-ca20-11eb-835f-b7b996c2db6d.png)

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: (yes / no) yes

### Technical details

* Python version: Run `python --version`. 3.9.5
* Django version: Look in your requirements.txt, or run `pip show django | grep Version`. 3.2.4
* Wagtail version: Look at the bottom of the Settings menu in the Wagtail admin, or run `pip show wagtail | grep Version:`. 2.13.1
* Browser version: You can use https://www.whatsmybrowser.org/ to find this out.  Chrome 91

**Comment by cnk** [comment]
I can confirm this issue exists (on Wagtail 2.13.2). One more detail, if the missing data is in the first StructBlock in the list, the "this field is required" is rendered correctly. But if the missing data is in a StructBlock later in the list, the error message is not rendered on the relevant field.

**Comment by rudeb0t** [comment]
I am having a similar issue to this one, except mine involves multi column blocks.

```python
class CustomStyledBlock(blocks.StructBlock):
    html_id = blocks.CharBlock(
        required=False, help_text=_("HTML id attribute " "value.")
    )
    heading = blocks.CharBlock(required=False)
    sub_heading = blocks.CharBlock(required=False)
    css_classes = blocks.CharBlock(required=False)


class RichTextContentBlock(CustomStyledBlock):
    text = blocks.RichTextBlock(editor="content-full")


class ColumnContentBlock(blocks.StreamBlock):
    hero_image = ImageBlock(template=f"{TEMPLATE_PREFIX}hero_image.html")
    content = RichTextContentBlock(template=f"{TEMPLATE_PREFIX}content.html")


class MultiColumnBlock(blocks.StreamBlock):
    two_columns = ColumnsBlock(
        ColumnContentBlock(),
        ratios=(1, 1),
        template=f"{TEMPLATE_PREFIX}/multi_column_block_2.html",
    )

    three_columns = ColumnsBlock(
        ColumnContentBlock(),
        ratios=(1, 1, 1),
        template=f"{TEMPLATE_PREFIX}/multi_column_block_3.html",
    )

    four_columns = ColumnsBlock(
        ColumnContentBlock(),
        ratios=(1, 1, 1, 1),
        template=f"{TEMPLATE_PREFIX}/multi_column_block_4.html",
    )


class CompositePage(Page):
    content = StreamField(
        [
            ('hero_image', ImageBlock(template=f"{TEMPLATE_PREFIX}hero_image.html")),
            ('content', RichTextContentBlock(template=f"{TEMPLATE_PREFIX}content.html")),
            ('multi_column_block', MultiColumnBlock(template=f"{TEMPLATE_PREFIX}multi_column_block.html"))
        ]
    )
```
This worked setup fine up until 2.12.5. I have hundreds of pages using this page model (multi column with rich text content) and upon upgrading to 2.13.1 editing pages and saving them results in validation error. I also found that if more than one column has a rich text block only the first column will have a rendered rich text editor. The rich text editor on the other columns do not show up.

**Comment by gasman** [comment]
@rudeb0t Are you using the `wagtailcolumnblocks` package? Apparently that still needs updating for Wagtail 2.13 compatibility: https://github.com/squareweave/wagtailcolumnblocks/issues/8

**Comment by rudeb0t** [comment]
@gasman Yes, I have version 2.2.0 installed.

**Comment by rudeb0t** [comment]
@gasman Looks like I am hitting this issue as well as the one in squareweave/wagtailcolumnblocks#8 because my "regular" nested `StructBlock` content is also having validation errors.

**Comment by gasman** [comment]
@rudeb0t The issue here is about validation errors not showing when they _should_ be shown - it sounds like what you're describing is the opposite. In any case, I can only investigate that further if you can provide a test case that doesn't depend on wagtailcolumnblocks - otherwise, the default assumption is that this is a bug in that package.

**Comment by rudeb0t** [comment]
@gasman I have edited my comment above to clarify. I have several other blocks derived from my example `CustomStyledBlock` above and these are nested inside a `StructBlock` subclass like the one showed by the OP. I am having this exact same issue with the validation errors not showing up in the fields. Also have to add that I do not use custom forms for any of the blocks, if that helps.

**Comment by cnk** [comment]
I transplanted a demonstration of the error I am seeing into a more basic Wagtail project. Please see https://github.com/cnk/wagtail-tutorial/tree/ErrorDisplayIssue%237248#round-3-errors-not-displayed-for-deeply-nested-blocks

**Comment by gasman** [comment]
It looks like the error in @cnk's test case is caused by a mismatch in the definition of `ListBlockValidationError` between the backend and frontend code - the backend code builds up the error structure as a list with the same length as the data, and error-free items represented with `None`:

https://github.com/wagtail/wagtail/blob/1750c01d053c2c42ab76a4b67f51412847bf64e7/wagtail/core/blocks/list_block.py#L82-L88

while the front-end code is expecting it to be a dict (where the keys are list indexes where the errors are present):

https://github.com/wagtail/wagtail/blob/1750c01d053c2c42ab76a4b67f51412847bf64e7/client/src/components/StreamField/blocks/ListBlock.js#L134-L139

As a result, the front-end code isn't expecting `blockErrors` to contain null values, and fails with a Javascript error on the first child block with no errors.

@jkevingutierrez's test case doesn't use ListBlock, so I suspect there are actually multiple bugs with the same outcome (JS errors occurring during the call to `setError` causing not all errors to be shown).

**Comment by gasman** [comment]
@jkevingutierrez I can't replicate the bug under 2.13.1 following your test case. Can you confirm that the bug definitely occurs with the combination of blocks you describe? (I notice a couple of syntax errors in your code as posted, so I suspect that what you've posted here isn't exactly the setup you tested with.)

Here's what I tried:

* From a fresh `wagtail start myproject` project, edit home/models.py as follows:

```
from django.db import models

from wagtail.admin.edit_handlers import StreamFieldPanel
from wagtail.core import blocks
from wagtail.core.fields import StreamField
from wagtail.core.models import Page


class CopyBlock(blocks.StructBlock):
    text = blocks.RichTextBlock(required=True)


class GridLayoutBlock(blocks.StructBlock):
    body = blocks.StreamBlock(
        [
            ('copy', CopyBlock())
        ]
    )


class HomePage(Page):
    pass


class CustomPage(Page):
    body = StreamField([
        ('grid', GridLayoutBlock())
    ])

    content_panels = Page.content_panels + [
        StreamFieldPanel('body'),
    ]
```

* Run `./manage.py makemigrations` , `./manage.py migrate` , `./manage.py createsuperuser`
* Log in to admin and create a page of type Custom Page
* Fill in the page title, add a Grid block to the body field and add a Copy block to that, enter some text into it and then Save Draft
* Delete the text from the rich text block and then Save Draft again
* All the error indicators (including the badge on the Content tab and the "This field is required." message against the field) are shown as expected

**Comment by gasman** [comment]
Just a little nudge on this...
I'm hoping to get the 2.13.3 release out early next week, which will include the ListBlock fix from #7295 which fixes @cnk's test case. I'm open to the possibility that there's a separate bug here that affects non-ListBlock setups, but without a reproducible test case I'm not able to confirm that.

If anyone's able to provide such a test case by Monday (or confirm that @jkevingutierrez's steps above do show the issue, given a slightly different setup to what I tried...) then there's a good chance that the fix will make it in to 2.13.3.

**Comment by gasman** [comment]
Closing for now - happy to reopen if more reproducible test cases surface that aren't addressed by #7295.
